ffi/derive/src/arg.rs

```diff
@@ -0,0 +1,509 @@
+        unimplemented!("Not needed as of yet")
+    }
+    fn visit_type_path(&mut self, node: &'ast syn::TypePath) {
+        let last_seg = node.path.segments.last().expect_or_abort("Defined");
+
+        let arg_name = self.name;
+        match last_seg.ident.to_string().as_str() {
+            "bool" => {
+                if self.is_input {
+                    self.ffi_type = Some(parse_quote! { u8 });
+                    self.src_to_ffi = Some(parse_quote! {let #arg_name = #arg_name as u8;});
+                    self.ffi_to_src = Some(parse_quote! {let #arg_name = #arg_name != 0;});
+                } else {
+                    self.ffi_type = Some(node.clone().into());
+                    self.src_to_ffi = Some(parse_quote! {();});
+                    self.ffi_to_src = Some(parse_quote! {();});
+                }
+            }
+            // TODO: Following two match arms are quite similar
+            "u8" | "u16" => {
+                self.ffi_type = Some(parse_quote! {u32});
+                self.src_to_ffi = Some(parse_quote! {let #arg_name = #arg_name as u32;});
+                self.ffi_to_src = Some(parse_quote! {
+                    let #arg_name = match #arg_name.try_into() {
+                        Err(err) => return iroha_ffi::FfiResult::ConversionError,
+                        Ok(item) => item,
+                    };
+                });
+            }
+            "i8" | "i16" => {
+                self.ffi_type = Some(parse_quote! {i32});
+                self.src_to_ffi = Some(parse_quote! {let #arg_name = #arg_name as i32;});
+                self.ffi_to_src = Some(parse_quote! {
+                    let #arg_name = match #arg_name.try_into() {
+                        Err(err) => return iroha_ffi::FfiResult::ConversionError,
+                        Ok(item) => item,
+                    };
+                });
+            }
+            "u32" | "i32" | "f32" | "f64" => {}
+            "Option" => {
+                let item = generic_arg_types(last_seg)[0];
+                self.visit_type_option(item);
+            }
+            // TODO: Revisit
+            "Result" => {
+                let args = generic_arg_types(last_seg);
+                let (ok_type, _) = (args[0], args[1]);
+
+                match ok_type {
+                    Type::Path(type_) => self.visit_type_path(type_),
+                    Type::Reference(type_) => self.visit_type_reference(type_),
+                    _ => abort!(ok_type, "Unsupported Result::Ok type"),
+                }
+            }
+            _ => {
+                if self.is_input {
+                    self.ffi_type = Some(parse_quote! { *const #node });
+                    self.src_to_ffi = Some(parse_quote! {
+                        let #arg_name: *const _ = #arg_name;
+                    });
+                    self.ffi_to_src = Some(parse_quote! {
+                        let #arg_name = Clone::clone(&*#arg_name);
```

Comment:
at the moment I'm doing an in depth clone of the given argument. This may be substituted with ownership transfer instead

Reply:
yeah, because ownership transfer is a bit difficult to do correct

Reply:
opened #2325